controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -176,6 +185,93 @@ public void initializeStreamWriters(final ClientFactory clientFactory,
     }
 
     /**
+     * Method to check retention policy and generate new periodic cuts and/or truncate stream at an existing stream cut.
+     * 
+     * @param scope scope
+     * @param stream stream
+     * @param policy retention policy
+     * @param recordingTime time of recording
+     * @param contextOpt operation context
+     * @return future.
+     */
+    public CompletableFuture<Void> autoRetention(final String scope, final String stream, final RetentionPolicy policy,
+                                                 final long recordingTime, final OperationContext contextOpt) {
+        final OperationContext context = contextOpt == null ? streamMetadataStore.createContext(scope, stream) : contextOpt;
+
+        return streamMetadataStore.getStreamCutsFromRetentionSet(scope, stream, context, executor)
+                .thenCompose(retentionSet -> {
```

Comment:
In which situation do we expect the set not to be a singleton? My initial expectation is that there is at any time at most one cut per stream, but this seems to indicate otherwise.

Reply:
basically retention set contains all stream cuts that we compute periodically.
say we create a stream on 1st jan. with a retention policy of 1 month retention.
now every half an hour we run a retention workflow for this stream.
this workflow computes a stream cut. So we have 48 stream cuts recorded per day till end of the month. 
Now on first feburary, we realize that we have reached a point where we need to trim such that we only keep last one month data. 
On 2nd feb we need to keep data till only 2nd jan. so we will have removed all stream cuts before 2nd jan.

So from this point on this is a rolling set every half an hour. we add a new stream cut to retention set and remove the oldest (we can skip to whichever streamcut maps to retention point). We truncate the stream at the oldest point. 

So in steady state in this example we will have 48 stream cuts stored in the retention set of this stream.

Reply:
It sounds we are trying to be conservative and compute + store stream cuts at a finer granularity than we might need. In this example, we don't really need that many stream cuts, so I'm wondering if can't adjust the granularity according to the policy of the stream.

Reply:
we have a cluster level configuration added for setting the frequency. 
But its not stream specific because that will make the configuration for application cumbersome.

Reply:
why can't we extract automatically from the stream configuration rather than the controller configuration? If the retention we set for a stream is for 30 days, why can't I set the frequency to close to 30 days? In fact, why can't I simply generate a new stream cut only upon a new truncation round?

Reply:
retention is "rolling". The stream with 1 month retention should have data worth last one month with some tolerance. 
That limit could be per stream too.. like the retention policy could look like following:
`RetentionPolicy {
time
truncationFrequency
}

But then we are putting the onus on customers to choose a truncation frequency and it can be arbitrarily large or small. 
If its large, we can truncate much later than desired. 
If its small, we waste a lot of compute cycles and resources on controller. 

The problem amplifies when we do it about retentionBySize. 
Its not obvious to the application developer what the frequency should be in such situation. 

With a reasonable periodicity chosen at controller we control the frequency at which we collect streamCuts per stream and perform truncation workflow per stream such that 
a) its not too costly to provide this automatic capability
b) we can limit the error in our retention to `less than frequency` for time based retention and
error limited to data written between two retention-workflow cycles.

Reply:
There is something I'm missing here, so let me go back to the basics. Say I create a stream on January 1st and that the retention is 30 days. When it hits January 31st, it should delete all data prior to January 1st, which is none, but for that it only needs a stream cut on January 1st, it doesn't need any stream cut in the middle. Following the example, on March 3rd, there is another truncation round, which needs a stream cut from January 31st.

My very basic question is why we need finer granularity of stream cuts given that all I need is a stream cut at the beginning of the period.

Reply:
Basically lets define the goal:
What does auto-retention with `1 month` mean.
It means we should try to keep data worth last one month.

Now there are two definitions involved here: 
1. retention duration : This is a value chosen by application and is per stream. 
2. frequency of collection: This is the rate at which we collect stream cuts and run truncation workflows. 
Frequency directly translates to degree of error we allow in retention point. 
So if frequency is 1 day, then for a `one month` retention, our last chosen truncation point can introduce an error of max 1 day. 
To understand this error better take the following example:

- On March 3, 0100 hours this means data till feb 1 1100 hours. 

- On March 3, 1100 hours, one month means feb 1 1100 hours.
But if we had collected stream cuts only at 0000 hours on each day, then error here at this point will be 11 hours.
For a 1 day retention cycle, this error will be maximum 23 hours 59 minutes 

- smaller granularity/ frequency of collection means smaller error. But it also means more resources on controller. 
- larger granularity/frequency of collection means larger error. 

IMO letting applications choose this frequency means we run the risk of them choosing a bad value either for us (small) or for themselves (large).
The frequency choice is still intuitive for `time based retention` but without knowing our internal implementation details, this choice for `size based retention` becomes harder. 

That is why cluster admin (or our supplied default) is a better choice for frequency. 

`Note: retention period is chosen by application developer. Frequency is a cluster level configuration`

Reply:
I'm not sure I understand this statement:

> Frequency directly translates to degree of error we allow in retention point.
> So if frequency is 1 day, then for a one month retention, our last chosen truncation point can introduce an error of max 1 day.

It seems to imply that the frequency of generating cuts has to be decoupled from the retention period, and under the assumption that they are decoupled, I agree that frequency and accuracy are at odds. But, the point here is that I think we can adapt the frequency of generating cuts according to the stream retention on an individual basis. If I understand it correctly, it would be a matter of setting this loop delay here: 

https://github.com/pravega/pravega/pull/2046/files#diff-62f7f286569f71c81a1f2c3d90b65733R115

Reply:
Implementation is not the issue. That's trivial.
So please give me an example of frequency as a function of retention period. 

Also as I said earlier, this function would become more interesting for size based retention.

Reply:
> It seems to imply that the frequency of generating cuts has to be decoupled from the retention period,
No. I am simply making these three points:
1. The error has a direct 1:1 correlation with the frequency of collection (no matter if the frequency is calculated or configured).

2. We dont have an obvious function for frequency which can be derived from user supplied retention policy (which can be size based or time based). 

3.  Small frequency has detrimental effect controller's resources, large frequencies increase the degree of error.  
We are better suited to choose frequency as opposed to application developer creating the stream

Reply:
My confusion is really basic, I don't understand why for time-based retention we need to talk about any frequency that is different from 1/p, where p is the retention period. For example, if the retention period is 30 days, then I really only need one new stream cut every 30 days. The generation of the cuts of course needs to be aligned with the retention, but otherwise, that's all we need, no?

Reply:
> For example, if the retention period is 30 days, then I really only need one new stream cut every 30 days. 

that way you have an error of 30 days in the system. (lets say all months have 30 days)
say you take a stream cut on 1st of every month. 
so on feb 30th, you cant use stream cut taken on feb 1. So you have data worth 50 days in the system when the application asked for data to be retained for only 30 days!

If we had taken a new stream cut every day, then we can ensure that we only keep the data for past 30 + 1 day